zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,373 @@
+    Preconditions.checkState(mostRecentLogNeededIndex >= 0);
+    Preconditions.checkState(mostRecentLogNeededIndex < logs.size());
+    Preconditions.checkState(oldestLogNeededIndex >= mostRecentLogNeededIndex);
+    Preconditions.checkState(snapNeededIndex >= 0);
+    Preconditions.checkState(snapNeededIndex < snaps.size());
+
+    BackupFileInfo snap = snaps.get(snapNeededIndex);
+    BackupFileInfo oldestLog = logs.get(oldestLogNeededIndex);
+    BackupFileInfo newestLog = logs.get(mostRecentLogNeededIndex);
+
+    if (oldestLog.getFileType() == BackupFileType.LOSTLOG) {
+      LOG.error("Could not find logs to make the snapshot '" + snap.getBackedUpFile()
+          + "' valid. Lost logs at " + logs.get(oldestLogNeededIndex).getZxidRange());
+      return false;
+    }
+
+    if (newestLog.getFileType() == BackupFileType.LOSTLOG) {
+      if (zxidToRestore == Long.MAX_VALUE && oldestLogNeededIndex != mostRecentLogNeededIndex) {
+        // When restoring to the latest, we can use the last valid log prior to lost log
+        // range.
+        mostRecentLogNeededIndex++;
+      } else {
+        LOG.error("Could not find logs to make the snapshot '" + snap.getBackedUpFile()
+            + "' valid. Lost logs at " + logs.get(mostRecentLogNeededIndex).getZxidRange() + ".");
+        return false;
+      }
+    }
+
+    Range<Long> fullRange = oldestLog.getZxidRange().span(newestLog.getZxidRange());
+
+    if (fullRange.lowerEndpoint() > snap.getZxidRange().lowerEndpoint()) {
+      LOG.error("Could not find logs to make snap '" + snap.getBackedUpFile()
+          + "' valid. Logs start at zxid " + ZxidUtils.zxidToString(fullRange.lowerEndpoint())
+          + ".");
+      return false;
+    }
+
+    if (fullRange.upperEndpoint() < snap.getZxidRange().upperEndpoint()) {
+      LOG.error("Could not find logs to make snap '" + snap.getBackedUpFile()
+          + "' valid. Logs end at zxid " + ZxidUtils.zxidToString(fullRange.upperEndpoint()) + ".");
+      return false;
+    }
+
+    if (zxidToRestore != Long.MAX_VALUE && fullRange.upperEndpoint() < zxidToRestore) {
+      LOG.error("Could not find logs to restore to zxid " + zxidToRestore + ". Logs end at zxid "
+          + ZxidUtils.zxidToString(fullRange.upperEndpoint()) + ".");
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Copy selected backup files from backup storage to a local restore temporary directory for further processing later
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @throws IOException
+   */
+  private void copyRestoredFilesToLocalTempDir(FileTxnSnapLog restoreTempSnapLog)
+      throws IOException {
+    // Copy backup files to local temp directory
+    for (BackupFileInfo backedupFile : filesToCopy) {
+      String standardFilename = backedupFile.getStandardFile().getName();
+      // Does not matter if it's dataDir or logDir since we use same path for these two directories
```

Comment:
What did you mean by this? This comment is confusing because I think it's possible that dataDir and logDir are different for certain systems? I think what you meant is that we just make them the same for the local temp directory FileSnapLog object for restoration. Let's clarify this.